### PR TITLE
Pack on a subset of blocks in a MeshData object

### PR DESCRIPTION
**Pull request by jdolence**
## PR Summary
This PR includes three things:
- It slightly expands the capability of `SparsePacks` by adding a new optional argument to `GetPack` that is a `std::vector<bool>` of flags that determine which blocks in a `MeshData` object actually get packed.
- A new `MetadataFlag` was added, `Metadata::AllocOnNewBlocks`, intended to force allocation of sparse fields whenever a new `MeshBlock` is created through AMR/load balancing.
- A refactor of the `DataCollection` add functionality so that when `Add` is called on a `MeshData` object, it appropriately calls `Add` on the block-level `DataCollection`s.  The benefit of this is that it can avoid having to always make entirely new `MeshBlockData` objects, instead just using the previously created objects when they're available.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/interface/mesh_data.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  pmy_mesh_ = src->GetParentPointer();
+  const int nblocks = src->NumBlocks();
+  block_data_.resize(nblocks);
+  for (int i = 0; i < nblocks; i++) {
+    block_data_[i] = pmy_mesh_->block_list[i]->meshblock_data.Add(
+        stage_name_, src->GetBlockData(i), names, shallow);
+  }
+}
+
+template class MeshData<Real>;
```
Comment (Yurlungur) [inline src/interface/mesh_data.cpp:34]: I *think* this should help with compile times.

File: src/interface/meshblock_data.cpp [hunk 2]
```diff
@@ -53,6 +53,12 @@ void MeshBlockData<T>::Initialize(
   for (auto const &q : resolved_packages->AllFields()) {
     AddField(q.first.base_name, q.second, q.first.sparse_id);
   }
+
+  Metadata::FlagCollection flags({Metadata::Sparse, Metadata::AllocOnNewBlocks});
+  auto vars = GetVariablesByFlag(flags);
+  for (auto &v : vars.vars()) {
+    AllocateSparse(v->label());
+  }
```
Comment (Yurlungur) [inline src/interface/meshblock_data.cpp]: Nice

File: src/interface/swarm_container.hpp [hunk 3]
```diff
@@ -43,7 +43,8 @@ class SwarmContainer {
   // Public Methods
   //-----------------
   // Constructor does nothing
-  SwarmContainer() {}
+  SwarmContainer() = default;
+  explicit SwarmContainer(const std::string &name) : swarm_name_(name) {}
```
Comment (Yurlungur) [inline src/interface/swarm_container.hpp:47]: Not that I object, but why this change?

File: src/interface/sparse_pack.hpp [hunk 4]
```diff
@@ -131,8 +131,8 @@ class SparsePack : public SparsePackBase {
     // accessed on device via instance of types in the type list Ts...
     // The pack will be created and accessible on the device
     template <class T>
-    SparsePack GetPack(T *pmd) const {
-      return SparsePack(SparsePackBase::GetPack(pmd, *this));
+    SparsePack GetPack(T *pmd, const std::vector<bool> &include_block = {}) const {
+      return SparsePack(SparsePackBase::GetPack(pmd, *this, include_block));
```
Comment (Yurlungur) [inline src/interface/sparse_pack.hpp:135]: Is the `include_block` functionality expected to be used in userspace? Or do you plan to have some mechanism for automatically choosing which blocks to include?

**Review by Yurlungur — COMMENTED** [review]
Should `AllocOnNewBlocks` maybe be called `ForceAllocOnNewBlocks`?

Should probably test that the `include_block` stuff works in the unit tests. Also missing changelog. ;)

**Review by Yurlungur — APPROVED** [review]
Just going to approve now, so I don't forget, pending tests, changelog, formatting, etc.

File: src/interface/sparse_pack.hpp (hunk 4 above)
Comment (jdolence) [reply src/interface/sparse_pack.hpp:135]: user space.  i think that's a lot more flexible for downstream codes.  i don't presume to know how someone may eventually use this, beyond the use case we have.  and for us, there is essentially zero burden to passing this in.

File: src/interface/swarm_container.hpp (hunk 3 above)
Comment (jdolence) [reply src/interface/swarm_container.hpp:47]: this was part of refactoring how the `Add` functionality works with `MeshData`.  all container like objects now have names which can be passed into their constructors.

File: src/interface/sparse_pack.hpp (hunk 4 above)
Comment (Yurlungur) [reply src/interface/sparse_pack.hpp:135]: Yeah ok fair enough

**Comment by jdolence** [comment]
> Should `AllocOnNewBlocks` maybe be called `ForceAllocOnNewBlocks`?
> 
> Should probably test that the `include_block` stuff works in the unit tests. Also missing changelog. ;)

I thought about `ForceAlloc...` vs `Alloc...`.  `Force` just feels really long and I decided it didn't really say anything new.  I don't feel strongly about this though, so I can change it if there's consensus around it.

**Comment by Yurlungur** [comment]
> I thought about `ForceAlloc...` vs `Alloc...`. `Force` just feels really long and I decided it didn't really say anything new. I don't feel strongly about this though, so I can change it if there's consensus around it.

IMO `Alloc...` is confusing as it implies that the default behavior would be to never allocate... whereas the real default behavior is to allocate sometimes.

File: src/interface/meshblock_data.cpp [hunk 5]
```diff
@@ -53,6 +53,12 @@ void MeshBlockData<T>::Initialize(
   for (auto const &q : resolved_packages->AllFields()) {
     AddField(q.first.base_name, q.second, q.first.sparse_id);
   }
+
+  Metadata::FlagCollection flags({Metadata::Sparse, Metadata::ForceAllocOnNewBlocks});
+  auto vars = GetVariablesByFlag(flags);
+  for (auto &v : vars.vars()) {
+    AllocateSparse(v->label());
+  }
```
Comment (lroberts36) [inline src/interface/meshblock_data.cpp:61]: Is the `Metadata::Sparse` required? Shouldn't that flag have to be set for `Metadata::ForceAllocOnNewBlocks` to be set?

File: src/mesh/amr_loadbalance.cpp [hunk 6]
```diff
@@ -134,7 +134,9 @@ bool TryRecvCoarseToFine(int lid_recv, int send_rank, const LogicalLocation &fin
             });
       }
     } else {
-      if (pmb->IsAllocated(var->label())) pmb->DeallocateSparse(var->label());
+      if (pmb->IsAllocated(var->label()) &&
+          !var->metadata().IsSet(Metadata::ForceAllocOnNewBlocks))
+        pmb->DeallocateSparse(var->label());
```
Comment (lroberts36) [inline src/mesh/amr_loadbalance.cpp:139]: Why no change for `TryRecvFineToCoarse`?

File: src/mesh/amr_loadbalance.cpp [hunk 7]
```diff
@@ -296,7 +298,9 @@ bool TryRecvSameToSame(int lid_recv, int send_rank, Variable<Real> *var, MeshBlo
       pmb->pmr->DereferenceCount() = counter_subview_h(0);
       var->dealloc_count = counter_subview_h(1);
     } else {
-      if (pmb->IsAllocated(var->label())) pmb->DeallocateSparse(var->label());
+      if (pmb->IsAllocated(var->label()) &&
+          !var->metadata().IsSet(Metadata::ForceAllocOnNewBlocks))
+        pmb->DeallocateSparse(var->label());
```
Comment (lroberts36) [inline src/mesh/amr_loadbalance.cpp:303]: Should there be an explicit allocation here if `Metadata::ForceAllocOnNewBlocks` is set?

**Review by lroberts36 — APPROVED** [review]
LGTM

File: src/mesh/amr_loadbalance.cpp (hunk 7 above)
Comment (jdolence) [reply src/mesh/amr_loadbalance.cpp:303]: That's handled in `MeshBlockData::Initialize`

File: src/interface/meshblock_data.cpp (hunk 5 above)
Comment (jdolence) [reply src/interface/meshblock_data.cpp:61]: As it's written right now, `Metadata::ForceAllocOnNewBlocks` has no effect on any variable that doesn't have `Metadata::Sparse`.  Those fields will always be allocated anyway, of course.  I could add some kind of check that'll throw if you set `Metadata::ForceAllocOnNewBlocks` for a dense variable, but I didn't because you still get the behavior you expect.

File: src/mesh/amr_loadbalance.cpp (hunk 6 above)
Comment (jdolence) [reply src/mesh/amr_loadbalance.cpp:139]: `DeallocateSparse` isn't called there.  Not sure if it should be.  It's actually not totally clear to me why it needs to be called in the other cases.  Maybe you know the answer to that?
